packages/next/server/app-render.tsx

```diff
@@ -537,54 +608,58 @@ export async function renderToHTML(
       }
     }
 
+    /**
+     * The React Component to render.
+     */
     const Component = layoutOrPageMod
       ? interopDefault(layoutOrPageMod)
       : undefined
 
+    // Handle dynamic segment params.
     const segmentParam = getDynamicParamFromSegment(segment)
+    /**
+     * Create object holding the parent params and current params, this is passed to getServerSideProps and getStaticProps.
+     */
     const currentParams =
       // Handle null case where dynamic param is optional
       segmentParam && segmentParam.value !== null
         ? {
             ...parentParams,
             [segmentParam.param]: segmentParam.value,
           }
-        : parentParams
-    const actualSegment = segmentParam
-      ? [segmentParam.param, segmentParam.treeValue]
-      : segment
+        : // Pass through parent params to children
+          parentParams
+    // Resolve the segment param
+    const actualSegment = segmentParam ? segmentParam.treeSegment : segment
```

Comment:
Found a small bug here that probably caused the random refreshing in some cases with dynamic routes.